### PR TITLE
feat: allow urls with authentication

### LINKED ISSUE #836: Make it possible to use Web UI with a remote IPFS API
Lots of people want to use Web UI to monitor and manage IPFS nodes hosted on a remote server rather than on their local machine. That is totally reasonable. 

But **it is not currently supported**. The IPFS API is not authenticated. It currently expects to be accessed by privileged users on the local machine, and should not be exposed to the internet. Once we have an authentication mechanism for the API we'll ensure that Web UI works with it.

Please see: https://github.com/ipfs/go-ipfs/issues/1532

In the meantime we'll make the subset of Web UI that can work against the read-only / gateway API that would be safe expose from a remote machine. See #835

A suitably motivated individual _could_ set up secure proxy + firewall or vpn to access their api port securely from a remote machine. Pervious iterations of Web UI were hardcoded to use an API on localhost, but v2 will do nothing special to prevent trying to use a remote api address. It's worth reiterating that its not a currently supported configuration, and we don't have the bandwidth to deal with support requests if it doesn't work for you. We will make this work in the future, in a secure way, once core supports it.

### Previous reports of this issue

- https://github.com/ipfs-shipyard/ipfs-webui/issues/637
- https://github.com/ipfs-shipyard/ipfs-webui/issues/628
- https://github.com/ipfs-shipyard/ipfs-webui/issues/594
- https://github.com/ipfs-shipyard/ipfs-webui/issues/591
- https://github.com/ipfs-shipyard/ipfs-webui/issues/95
- https://discuss.ipfs.io/t/how-can-i-enable-remote-connection-to-webui/698/3

### PR DESCRIPTION
**Pull request by hacdias**
This should fix https://github.com/ipfs-shipyard/ipfs-webui/issues/836 by allowing to pass custom URLs.

I was not able to try it due to some CORS configuration with Basic Auth I couldn't manage to get working right now. Will try later. @lidel could you take a look at this to see if it looks OK or even try it if you've got time?

License: MIT
Signed-off-by: Henrique Dias <hacdias@gmail.com>

### CODE REVIEW

**Review by lidel — COMMENTED** [review]
I will try to test this against real basic auth setup in a spare moment today/tomorrow, for now quick feedback below.

File: src/index.js [hunk 1]
```diff
@@ -182,6 +185,20 @@ function isMultiaddress (addr) {
   }
 }
 
+function isURL (addr) {
+  if (addr === null || addr === undefined || typeof addr === 'undefined') {
```
[the remark this answers is not in the document]
Comment (hacdias) [reply src/index.js]: True!

**Comment by hacdias** [comment]
Thanks @lidel!

**Review by lidel — CHANGES_REQUESTED** [review]
## Summary

So this PR works as expected :+1:

Remaining work in this PR:

- [ ] `src/js-ipfs-api/index.test.js` **needs tests** for API specified as URL (with and without basic auth)

There are bugs in WebUI that need to be addressed upstream 
(it does not support object, just string with multiaddr, details below)

## HTTP Proxy for testing this

Ok, there are many moving pieces, including Origin validation at go-ipfs, so I've wrote a custom http proxy to make it easier to test:

```js
#!/usr/bin/env node

const http = require('http')
const httpProxy = require('http-proxy')
const basicAuth = require('basic-auth')

// ============================================================
// CONFIG
// ============================================================
const port = 5403 // port on which we want to enable basic auth
const target = 'http://127.0.0.1:5001' // where  /api/v0/ lives
// basic auth:
const user = 'user'
const password = 'pass'
// ============================================================

const proxy = httpProxy.createProxyServer()

proxy.on('proxyReq', (proxyReq, req, res, options) => {
  // to make this poc easier, swap Origin so internal check of go-ipfs does get triggered (403 Forbidden on Origin mismatch)
  proxyReq.setHeader('Origin', target);
  proxyReq.setHeader('Referer', target);
  proxyReq.setHeader('Host', new URL(target).host);
})

http.createServer((req, res) => {
  console.log(`${req.method}\t\t${req.url}`)

  res.oldWriteHead = res.writeHead
  res.writeHead = function(statusCode, headers) {
    // hardcoded liberal CORS for easier testing
    res.setHeader('Access-Control-Allow-Origin', '*')
    // usual suspects + 'authorization' header
    res.setHeader('Access-Control-Allow-Headers', 'X-Stream-Output, X-Chunked-Output, X-Content-Length, authorization')
    res.oldWriteHead(statusCode)
  }

  const auth = basicAuth(req)

  const preflight = req.method === 'OPTIONS' // need preflight working
  if (!preflight && !(auth && auth.name === user && auth.pass === password)) {
    res.writeHead(401, {
      'WWW-Authenticate': 'Basic realm="IPFS API"'
    })
    res.end('Access denied')
  } else {
    proxy.web(req, res, { target })
  }
}).listen(port)
```

**TL;DR:** You run it via `node ./proxy.js`, then confirm it works via `curl http://user:pass@127.0.0.1:5403/api/v0/id` 

It should save us a TON of time with setting CORS headers.


## Testing WebUI with this PR


[connect-deps](https://www.npmjs.com/package/connect-deps) makes it easy to test webui with this PR: (in your local webui repo) `connect-deps link ../ipfs-redux-bundle/`, then `connect-deps connect -w`


## Test result 

Clicking on "Submit" in WebUI did nothing:

> 

So instead I set `ipfsApi` hint for WebUI via console:

```js
localStorage.setItem('ipfsApi','http://user:pass@127.0.0.1:5403')
```

console-setup-2019-11-01--11-56-52


Reloaded WebUI and `ipfs-redux-bundle` from this PR picked URL up correctly:



I was able to open `http://127.0.0.1:3000/#/files` and browse files using HTTP API over  Basic Auth.

## Bugs in WebUI

Found some bugs in WebUI tho (writing them down below, but need to be addressed upstream):

1. I was unable to set URL via `http://localhost:3000/#/welcome`, Submit button does nothing, I had to set it via console like described above.

2. <del> When set via console, the Connection screen at `http://localhost:3000/#/welcome` displays broken value for URL (looks like invalid string serialization):</del> fixed in latest version of this PR
  

3. Status page at `http://127.0.0.1:3000/#/files` crash because it expects multiaddr:
   url-multiaddr-err-2019-11-01--11-48-26

**Comment by hacdias** [comment]
@lidel I moved the checks and conversion of the api address to the `js-ipfs-api` directory in order to keep the `apiAddress` as a string all the times which might be easier for editing and submitting through the web ui. In addition, I added two tests.

Thanks for that proxy!!!!!!

File: src/utils.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
+'use strict'
+/* eslint-env browser, webextensions */
+
+const multiaddr = require('multiaddr')
+
+function isMultiaddress (addr) {
+  if (addr === null || addr === undefined || typeof addr === 'undefined') {
+    return false
+  }
+
```
Comment (lidel) [inline src/utils.js]: sneaked in during refactoring, can be removed, try/catch below is enough

**Review by lidel — APPROVED** [review]
LGTM! Just address small ask below and ok to ship it.

FYSA this fixed WebUI Bug (2) (URL set via console now looks ok), remaining bugs need to be addressed upstream, in against WebUI which switches to new ipfs-redux-bundle

File: src/utils.js (hunk 2 above)
Comment (hacdias) [reply src/utils.js]: This is on `isMultiaddr` and not on `isUrl`. Our `multiaddr` does not throw with undefined nor null. I'll leave this check, but the one on `isUrl` was removed.

**Comment by lidel** [comment]
Reintroducing this in https://github.com/ipfs-shipyard/ipfs-webui/pull/1613
